### ISSUE TITLE
Fields named "default" do not generate the expected API.

### ISSUE THREAD

**Issue by mjball** [immutables/immutables#431]
If I have this immutable type definition:

```java
@HubSpotStyle
@Value.Immutable
public interface AvatarIF {
  boolean getDefault();
  String getUrl();
}
```

where `@HubSpotStyle` is our style annotation containing the following:

```java
@Target({ ElementType.PACKAGE, ElementType.TYPE})
@Retention(RetentionPolicy.CLASS) // Make it class retention for incremental compilation
@JsonSerialize
@Value.Style(
    get = {"is*", "get*"},
    init = "set*",
    typeAbstract = {"Abstract*", "*IF"},
    typeImmutable = "*",
    throwForInvalidImmutableState = InvalidImmutableStateException.class,
    optionalAcceptNullable = true,
    forceJacksonPropertyNames = false,
    visibility = ImplementationVisibility.SAME)
public @interface HubSpotStyle {}
```

Immutables 2.2.10 generates the following code:

```java
package com.hubspot.avatars.core.primitives;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import javax.annotation.Generated;
import javax.annotation.Nullable;
import javax.annotation.ParametersAreNonnullByDefault;
import javax.annotation.concurrent.Immutable;
import javax.annotation.concurrent.NotThreadSafe;

import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.hubspot.immutables.exceptions.InvalidImmutableStateException;

import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;

/**
 * Immutable implementation of {@link AvatarIF}.
 * <p>
 * Use the builder to create immutable instances:
 * {@code Avatar.builder()}.
 */
@SuppressWarnings("all")
@SuppressFBWarnings
@ParametersAreNonnullByDefault
@Generated({"Immutables.generator", "AvatarIF"})
@Immutable
public final class Avatar implements AvatarIF {
  private final boolean getDefault;
  private final String url;

  private Avatar(boolean getDefault, String url) {
    this.getDefault = getDefault;
    this.url = url;
  }

  /**
   * @return The value of the {@code getDefault} attribute
   */
  @JsonProperty
  @Override
  public boolean getDefault() {
    return getDefault;
  }

  /**
   * @return The value of the {@code url} attribute
   */
  @JsonProperty
  @Override
  public String getUrl() {
    return url;
  }

  /**
   * Copy the current immutable object by setting a value for the {@link AvatarIF#getDefault() getDefault} attribute.
   * A value equality check is used to prevent copying of the same value by returning {@code this}.
   * @param getDefault A new value for getDefault
   * @return A modified copy of the {@code this} object
   */
  public final Avatar withGetDefault(boolean getDefault) {
    if (this.getDefault == getDefault) return this;
    return new Avatar(getDefault, this.url);
  }

  /**
   * Copy the current immutable object by setting a value for the {@link AvatarIF#getUrl() url} attribute.
   * An equals check used to prevent copying of the same value by returning {@code this}.
   * @param url A new value for url
   * @return A modified copy of the {@code this} object
   */
  public final Avatar withUrl(String url) {
    if (this.url.equals(url)) return this;
    String newValue = Objects.requireNonNull(url, "url");
    return new Avatar(this.getDefault, newValue);
  }

  /**
   * This instance is equal to all instances of {@code Avatar} that have equal attribute values.
   * @return {@code true} if {@code this} is equal to {@code another} instance
   */
  @Override
  public boolean equals(@Nullable Object another) {
    if (this == another) return true;
    return another instanceof Avatar
        && equalTo((Avatar) another);
  }

  private boolean equalTo(Avatar another) {
    return getDefault == another.getDefault
        && url.equals(another.url);
  }

  /**
   * Computes a hash code from attributes: {@code getDefault}, {@code url}.
   * @return hashCode value
   */
  @Override
  public int hashCode() {
    int h = 31;
    h = h * 17 + (getDefault ? 1231 : 1237);
    h = h * 17 + url.hashCode();
    return h;
  }

  /**
   * Prints the immutable value {@code Avatar} with attribute values.
   * @return A string representation of the value
   */
  @Override
  public String toString() {
    return "Avatar{"
        + "getDefault=" + getDefault
        + ", url=" + url
        + "}";
  }

  /**
   * Utility type used to correctly read immutable object from JSON representation.
   * @deprecated Do not use this type directly, it exists only for the <em>Jackson</em>-binding infrastructure
   */
  @Deprecated
  @JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.NONE)
  static final class Json implements AvatarIF {
    boolean getDefault;
    boolean getDefaultIsSet;
    @Nullable String url;
    @JsonProperty
    public void setGetDefault(boolean getDefault) {
      this.getDefault = getDefault;
      this.getDefaultIsSet = true;
    }
    @JsonProperty
    public void setUrl(String url) {
      this.url = url;
    }
    @Override
    public boolean getDefault() { throw new UnsupportedOperationException(); }
    @Override
    public String getUrl() { throw new UnsupportedOperationException(); }
  }

  /**
   * @param json A JSON-bindable data structure
   * @return An immutable value type
   * @deprecated Do not use this method directly, it exists only for the <em>Jackson</em>-binding infrastructure
   */
  @Deprecated
  @JsonCreator
  static Avatar fromJson(Json json) {
    Avatar.Builder builder = Avatar.builder();
    if (json.getDefaultIsSet) {
      builder.setGetDefault(json.getDefault);
    }
    if (json.url != null) {
      builder.setUrl(json.url);
    }
    return builder.build();
  }

  /**
   * Creates an immutable copy of a {@link AvatarIF} value.
   * Uses accessors to get values to initialize the new immutable instance.
   * If an instance is already immutable, it is returned as is.
   * @param instance The instance to copy
   * @return A copied immutable Avatar instance
   */
  public static Avatar copyOf(AvatarIF instance) {
    if (instance instanceof Avatar) {
      return (Avatar) instance;
    }
    return Avatar.builder()
        .from(instance)
        .build();
  }

  /**
   * Creates a builder for {@link Avatar Avatar}.
   * @return A new Avatar builder
   */
  public static Avatar.Builder builder() {
    return new Avatar.Builder();
  }

  /**
   * Builds instances of type {@link Avatar Avatar}.
   * Initialize attributes and then invoke the {@link #build()} method to create an
   * immutable instance.
   * <p><em>{@code Builder} is not thread-safe and generally should not be stored in a field or collection,
   * but instead used immediately to create instances.</em>
   */
  @NotThreadSafe
  public static final class Builder {
    private static final long INIT_BIT_GET_DEFAULT = 0x1L;
    private static final long INIT_BIT_URL = 0x2L;
    private long initBits = 0x3L;

    private boolean getDefault;
    private @Nullable String url;

    private Builder() {
    }

    /**
     * Fill a builder with attribute values from the provided {@code AvatarIF} instance.
     * Regular attribute values will be replaced with those from the given instance.
     * Absent optional values will not replace present values.
     * @param instance The instance from which to copy values
     * @return {@code this} builder for use in a chained invocation
     */
    public final Builder from(AvatarIF instance) {
      Objects.requireNonNull(instance, "instance");
      setGetDefault(instance.getDefault());
      setUrl(instance.getUrl());
      return this;
    }

    /**
     * Initializes the value for the {@link AvatarIF#getDefault() getDefault} attribute.
     * @param getDefault The value for getDefault 
     * @return {@code this} builder for use in a chained invocation
     */
    public final Builder setGetDefault(boolean getDefault) {
      this.getDefault = getDefault;
      initBits &= ~INIT_BIT_GET_DEFAULT;
      return this;
    }

    /**
     * Initializes the value for the {@link AvatarIF#getUrl() url} attribute.
     * @param url The value for url 
     * @return {@code this} builder for use in a chained invocation
     */
    public final Builder setUrl(String url) {
      this.url = Objects.requireNonNull(url, "url");
      initBits &= ~INIT_BIT_URL;
      return this;
    }

    /**
     * Builds a new {@link Avatar Avatar}.
     * @return An immutable instance of Avatar
     * @throws com.hubspot.immutables.exceptions.InvalidImmutableStateException if any required attributes are missing
     */
    public Avatar build() throws InvalidImmutableStateException {
      checkRequiredAttributes();
      return new Avatar(getDefault, url);
    }

    private boolean getDefaultIsSet() {
      return (initBits & INIT_BIT_GET_DEFAULT) == 0;
    }

    private boolean urlIsSet() {
      return (initBits & INIT_BIT_URL) == 0;
    }

    private void checkRequiredAttributes() throws InvalidImmutableStateException {
      if (initBits != 0) {
        throw new InvalidImmutableStateException(formatRequiredAttributesMessage());
      }
    }

    private String formatRequiredAttributesMessage() {
      List<String> attributes = new ArrayList<String>();
      if (!getDefaultIsSet()) attributes.add("getDefault");
      if (!urlIsSet()) attributes.add("url");
      return "Cannot build Avatar, some of required attributes are not set " + attributes;
    }
  }
}
```

I understand that `default` is a reserved keyword, so I do not expect that the field or parameters use that name. However, the public API this produces is incorrect, and prevents Jackson deserialization from working properly. Note the generated `Builder#setGetDefault()` method name. That should be `setDefault()`. Similar issues exist if I use `isDefault()` rather than `getDefault()`.

Similar issues also present when using other fields with keyword-like names, like `getLong()` for example.

What's the right course of action to fix this?

**Comment by elucash** [comment]
Yes, I think checks should be smarter in that "default" should be substituted with "getDefault", but derived "setDefault" should not become "setGetDefault" etc

**Comment by mjball** [comment]
It would probably be fine for internal field names to be prefixed or suffixed anyway, like `_default` for example.

**Comment by elucash** [comment]
changes released in 2.3. While I had to resort to not perfect solution, but it should be better now, for `isDefault`, `setDefault` will be generated, and JSON name would be `"default"`. Let us know if it does not work as expected. Thank you!
